### PR TITLE
feat: allow user to request membership

### LINKED ISSUE #1423: Request access to item
- user should be able to request access to given item
- admins can manage requests (set membership or delete)

### PR DESCRIPTION
**Pull request by pyphilia**
In this PR I tried to simplify a bit the code, by removing the use of the `ItemLoginAuthorization` black box.

It also has:
- button to request access to an item
- table to manage requests

TODO
- [x] update for enroll process

close #1423

### CODE REVIEW

**Comment by pyphilia** [comment]
Screenshot 2024-08-30 at 16 01 44

**Review by spaenleh — DISMISSED** [review]
Looks good to me ! I have added a few comments to improve some parts. Feel free to put them as issues if they are too large to handle in this PR. 

Great work on the refactoring of the code ! 💪

File: cypress/e2e/item/authorization/itemLogin/itemLogin.cy.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+      cy.visit(buildItemPath(item.id));
+      checkItemLoginScreenLayout(item.itemLoginSchema.type);
+      fillItemLoginScreenLayout({
+        username: 'username',
+        password: 'password',
+      });
+      cy.wait('@postItemLogin');
+
+      // use username to check no member id is incorrectly sent
```
Comment (ReidyT) [inline cypress/e2e/item/authorization/itemLogin/itemLogin.cy.ts:78]: Same question here

File: cypress/e2e/item/authorization/itemLogin/itemLogin.cy.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    );
+    cy.setUpApi({ items: [item], currentMember: MEMBERS.BOB });
+    cy.visit(buildItemPath(item.id));
+
+    // avoid to detect intermediate screens because of loading
+    // to remove when requests loading time is properly managed
+    cy.wait(ITEM_LOGIN_PAUSE);
+
+    // show enroll button
+    cy.click('iojwlmke');
```
Comment (ReidyT) [inline cypress/e2e/item/authorization/itemLogin/itemLogin.cy.ts:128]: I agree, it's not clear ^^

File: cypress/e2e/item/authorization/itemLogin/itemLoginSetting.cy.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+      expect(el.parent().html()).to.contain('disabled');
+    });
+  }
+};
+
+const editItemLoginSetting = (mode: string) => {
+  cy.get(`#${SHARE_ITEM_PSEUDONYMIZED_SCHEMA_ID}`).click();
+  cy.get(`li[data-value="${mode}"]`).click();
+  cy.wait('@putItemLoginSchema').then(({ request: { body } }) => {
+    expect(body?.type).to.equal(ItemLoginSchemaType.UsernameAndPassword);
```
Comment (ReidyT) [inline cypress/e2e/item/authorization/itemLogin/itemLoginSetting.cy.ts:46]: Shouldn't the body type be equal to the new mode?

File: cypress/e2e/item/authorization/itemLogin/itemLoginSetting.cy.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import { ITEM_LOGIN_PAUSE } from '../../../../support/constants';
+
+const checkItemLoginSetting = ({
+  isEnabled,
+  mode,
+  disabled = false,
+}: {
+  isEnabled: boolean;
+  mode: string;
+  disabled?: boolean;
```
Comment (ReidyT) [inline cypress/e2e/item/authorization/itemLogin/itemLoginSetting.cy.ts:26]: What is the difference between `isEnabled = true` and `disabled = false`? To me, the two parameters represent the same thing 😅.

File: src/components/pages/item/accessWrapper/RequestAccessContent.tsx [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    <Stack
+      direction="column"
+      justifyContent="center"
+      alignItems="center"
+      height="100%"
+      gap={2}
+    >
+      <Lock size={40} />
+      <Typography variant="h3">
+        {translateBuilder('Request access to this item')}
```
Comment (ReidyT) [inline src/components/pages/item/accessWrapper/RequestAccessContent.tsx:67]: Use const here too

File: src/components/pages/item/accessWrapper/RequestAccessContent.tsx [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        disabled={isSuccess}
+        loading={isLoading}
+        endIcon={isSuccess ? <Check /> : null}
+        onClick={async () => {
+          await requestMembership({ id: itemId });
+        }}
+      >
+        {isSuccess
+          ? translateBuilder('request sent')
+          : translateBuilder('Request access')}
```
Comment (ReidyT) [inline src/components/pages/item/accessWrapper/RequestAccessContent.tsx:81]: Same here

**Review by ReidyT — CHANGES_REQUESTED** [review]
Thanks for this PR, it's a great refactor 🙌 I agree with what Basile said and I also let a few comments. Except for that, it looks good to me 🤠

File: src/components/item/sharing/membershipTable/EditPermissionButton.tsx [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+                allowDowngrade={allowDowngrade}
+              />
+            </Stack>
+          </Stack>
+        </DialogContent>
+        <DialogActions>
+          <Button variant="text" onClick={closeModal}>
+            {translateCommon(COMMON.CANCEL_BUTTON)}
+          </Button>
+          <Button
```
[the remark this answers is not in the document]
Comment (spaenleh) [reply src/components/item/sharing/membershipTable/EditPermissionButton.tsx:92]: Ok so it is good !

**Pull request merged**